File: src/main/java/org/tensorflow/utils/SessionRunner.java
package org.tensorflow.utils;

import org.tensorflow.Operand;
import org.tensorflow.Session;
import org.tensorflow.Tensor;
import org.tensorflow.types.family.TType;

import java.util.List;

public class SessionRunner {
  private Session.Runner runner;

  public SessionRunner(Session.Runner runner) {
    this.runner = runner;
  }

  public SessionRunner addTarget(String operation) {
    this.runner.addTarget(operation);
    return this;
  }

  public SessionRunner addTarget(Operand operand) {
    this.runner.addTarget(operand);
    return this;
  }

  public SessionRunner addTargets(Operand<?>... operands) {
    for (Operand<?> operand : operands) {
      this.runner.addTarget(operand);
    }
    return this;
  }

  public <T extends TType> SessionRunner addTargets(List<Operand<T>> operands) {
    for (Operand<?> operand : operands) {
      this.runner.addTarget(operand);
    }
    return this;
  }

  public SessionRunner addTargets(String... targets) {
    for (String target : targets) {
      this.runner.addTarget(target);
    }

    return this;
  }

  public SessionRunner feed(Tensor[] tensors, Operand[] ops) {
    for (Pair<Tensor, Operand> pairs :
        (Iterable<Pair<Tensor, Operand>>) () -> Pair.zip(tensors, ops)) {
      this.runner.feed(pairs.second().asOutput(), pairs.first());
    }
    return this;
  }

  public SessionRunner fetch(String operation) {
    this.runner.fetch(operation);
    return this;
  }

  public SessionRunner fetch(Operand operand) {
    this.runner.fetch(operand);
    return this;
  }

  public SessionRunner fetch(String... operations) {
    for (String operation : operations) {
      this.runner.fetch(operation);
    }
    return this;
  }

  public SessionRunner fetch(Operand<?>... operands) {
    for (Operand<?> operand : operands) {
      this.runner.fetch(operand);
    }
    return this;
  }

  public <T extends TType> SessionRunner fetch(List<Operand<T>> operands) {
    for (Operand<?> operand : operands) {
      this.runner.fetch(operand);
    }
    return this;
  }

  public List<Tensor/*<?>*/> run() {
    return this.runner.run();
  }
}

File: src/main/scala/org/tensorflow/utils/SessionRunner.scala
package org.tensorflow.utils

import org.tensorflow.Operand
import org.tensorflow.Session
import org.tensorflow.Tensor
import org.tensorflow.types.family.TType

import java.{util => ju}

class SessionRunner(val runner: Session#Runner) {
  def addTarget(operation: String): SessionRunner = {
    this.runner.addTarget(operation)
    this
  }

  def addTarget(operand: Operand[_]): SessionRunner = {
    this.runner.addTarget(operand)
    this
  }

  def addTargets(operands: Operand[_]*): SessionRunner = {
    for (operand <- operands) {
      this.runner.addTarget(operand)
    }
    this
  }

  def addTargets[T <: TType](operands: ju.List[Operand[T]]): SessionRunner = {
    operands.forEach { operand =>
      this.runner.addTarget(operand)
    }
    this
  }

  def addTargetsStr(targets: String*): SessionRunner = {
    for (target <- targets) {
      this.runner.addTarget(target)
    }
    this
  }

  def feed(tensors: Array[Tensor], ops: Array[Operand[_]]): SessionRunner = {
    for (pairs <- tensors zip ops) {
      this.runner.feed(pairs._2.asOutput, pairs._1)
    }
    this
  }

  def fetch(operation: String): SessionRunner = {
    this.runner.fetch(operation)
    this
  }

  def fetch(operand: Operand[_]): SessionRunner = {
    this.runner.fetch(operand)
    this
  }

  def fetchSeqStr(operations: String*): SessionRunner = {
    for (operation <- operations) {
      this.runner.fetch(operation)
    }
    this
  }

  def fetchSeq(operands: Operand[_]*): SessionRunner = {
    for (operand <- operands) {
      this.runner.fetch(operand)
    }
    this
  }

  def fetch[T <: TType](operands: ju.List[Operand[T]]): SessionRunner = {
    operands.forEach { operand =>
      this.runner.fetch(operand)
    }
    this
  }

  def run: ju.List[Tensor] = this.runner.run
}
